timesync/clock.go

```diff
@@ -0,0 +1,121 @@
+package timesync
+
+import (
+	"github.com/spacemeshos/go-spacemesh/mesh"
+	"sync"
+	"time"
+)
+
+type LayerTimer chan mesh.LayerID
+
+type Ticker struct {
+	subscribes []LayerTimer
+	currentLayer mesh.LayerID
+	m sync.Mutex
+	tickInterval time.Duration
+	startEpoch time.Time
+	time Clock
+	stop chan struct{}
+}
+
+type Clock interface {
+	Now() time.Time
+}
+
+func NewTicker(time Clock, tickInterval time.Duration, startEpoch time.Time) *Ticker {
+	return &Ticker{
+		subscribes: make([]LayerTimer,0,0),
+		currentLayer: 0,
+		tickInterval:tickInterval,
+		startEpoch:startEpoch,
+		time : time,
+		stop : make(chan struct{}),
+	}
+}
+
+func (t *Ticker) Start(){
+	go t.StartClock()
+}
+
+func (t *Ticker) Stop(){
+	close(t.stop)
+}
+
+func (t *Ticker) notifyOnTick(){
+	t.m.Lock()
```

Comment:
any reason to not use a read/write lock ? not critical since this will happen only on ticks anyway